Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1333 @@
+
+This tutorial was written using GHC version 9.0.2, Stack resolver LTS-19.8, Alex version 3.2.7.1, and Happy version 1.20.0.
+
+## Our grammar: MiniML
+
+We've discussed the grammar in the previous article, but let's reintroduce it here.
+MiniML (read as "minimal") will be a small grammar based on ML.
+By the end of this tutorial, you will be able to parse programs such as the following:
+
+```sml
+let the_answer : int =
+  let a = 20 in
+  let b = 1 in
+  let c = 2 in
+  a * c + b * c
+
+let main (unit : ()) : () =
+  print ("The answer is: " + the_answer)
+```
+
+Some of the features that this tutorial will support are:
+* Variables, constants, function declarations, function applications, local bindings, and operators.
+* Some simple types, as well as arrow types (functions).
+* Conditional expressions: if-then-else and if-then.
+* Block comments (added in part 1).
+* The reader will be invited in the proposed exercises to extend the language with patterns, list access, and pattern matching. The exercises in the first part also introduced rational numbers and line comments.
+
+## Comparison with parser combinators
+
+<!--
+* [X] refer to parser combinators article
+* [X] talk about the pros and cons
+-->
+
+In a previous tutorial, we discussed implementing and using [Parser Combinators in Haskell](https://serokell.io/blog/parser-combinators-in-haskell).
+In the article, we discussed some pros and cons of LL parsers compared to LR parsers, which we will reiterate and elaborate on.
+
+<!--
+TODO: not sure how correct is the theory here, a more in-depth review here would be nice
+-->
+
+Haskell libraries such as Megaparsec work best on LL(1) or LL(∞) grammars and they're implemented as recursive descent parsers with backtracking.
+As such, caveats like not being able to handle left-recursion also apply to them, besides that their performance can exponentially degrade due to backtracking.
+LL here means **L**eft-to-right parsing, **L**eftmost derivation, as they scan strings from left to right and build a tree deriving them from the left (see below).
+
+The number between the parentheses indicates how many look-ahead tokens the parser can see.
+A parser with a look-ahead of `k` can make decisions by looking `k` tokens (or characters) ahead.
+A look-ahead of `k=1` is often chosen for better performance, but popular parser combinatrs libraries offer an option to use an arbitrary `k` with the `try` function. A LL(k) parser is often enough to parse most grammars in programming languages.
+
+On the other hand, Happy is an LR(1)\* parser generator, where the R means **R**ightmost derivation in reverse.
+
+<details>
+<summary>* A note on Happy</summary>
+* Happy is actually an LALR(1) (Look-Ahead LR) parser generator, which is a more performant albeit less powerful variant of an LR(1) parser; see <a href=https://monlih.github.io/happy-docs/#_sec_conflict_tips>section 8.4.1</a> of the Happy User Guide.
```

Comment:
I think I'd prefer to be told that it is a LALR(1) parser outright. Otherwise, it might be a bit confusing. 

If you rename this section "comparison with Megaparsec et al", a full explanation will be right in place.

Reply:
Changed it.

Could you elaborate on your second point? As per your suggestion below, I renamed it to "Happy vs. Megaparsec". But I don't understand if you meant that this would provide a full explanation of how it compares to Megaparsec, or if you want me to provide more points about it.

Reply:
I think that what you have now is good. 👍